File: 0743-network-delay-time/0743-network-delay-time.go
func networkDelayTime(times [][]int, n int, k int) int {
    // we can use djikstra here, to calculate the shortest distance from node k to all other nodes
    // dist = djikstra(k, graph)
    // answer = max(dist)
    adjList := make([][]Vertex, n)
    for _, time := range times { // --> this is O(E), E = number of edges
        s, d, cost := time[0]-1, time[1]-1, time[2]
        adjList[s] = append(adjList[s], Vertex{
            d, cost,
        })
    }

    dist := djikstra(k-1, adjList)
    var res = 0
    for i:=0;i<n;i++{ // --> this is O(V), where V = number of nodes
        if dist[i] == math.MaxInt {
            return -1
        }

        if res < dist[i] {
            res = dist[i]
        }
    }
    return res
}

func djikstra(source int, adjList [][]Vertex) []int {
    n := len(adjList)
    dist := make([]int, n)
    visited := make([]bool, n)
    for i:=0;i<n;i++{ // --> O(V)
        dist[i] = math.MaxInt
    }
    dist[source] = 0

    pq := NewMinHeap[Vertex](LessFunc)
    heap.Push(pq, Vertex{source, 0})

    for pq.Len() > 0 { // --> this is we try to iterate all of the graph, the edges dynamically generated from each iteration from the binaryHeap
        source := heap.Pop(pq).(Vertex) // -> this is O(log V)

        if visited[source.node]{
            continue
        }

        visited[source.node] = true
        
        // visit all the neighborhood
        for _, next := range adjList[source.node] { // for each node, we try to get all the edges to inside the heap. This is O(E)
            if !visited[next.node] && source.cost + next.cost < dist[next.node] {
                dist[next.node] = source.cost + next.cost
                heap.Push(pq, Vertex{ // -> O(Log V)
                    next.node,
                    dist[next.node],
                })
            } 
        }
    }
    // in summary it's gonna be O(V + E log V)

    return dist
}


type Vertex struct {
    node int
    cost int
}

func LessFunc(a, b Vertex) bool {
    return a.cost < b.cost
}

type MinHeap[T any] struct {
    data []T
    lessFunc func(a, b T)bool
}

func NewMinHeap[T any](lessFunc func(a, b T) bool, data ...T) *MinHeap[T] {
    res := MinHeap[T]{
        lessFunc: lessFunc,
        data: make([]T, len(data)),
    }

    if len(data) > 0 {
        copy(res.data, data)
        heap.Init(&res)
    }
    return &res
    
}

func (m MinHeap[T]) Len() int{
    return len(m.data)
}

func (m MinHeap[T]) Less(a, b int) bool {
    return m.lessFunc(m.data[a], m.data[b])
}

func (m MinHeap[T]) Swap(a, b int) {
    m.data[a], m.data[b] = m.data[b], m.data[a]
}

func (m *MinHeap[T]) Pop() any {
    l := m.Len()
    item := m.data[l-1]
    m.data = m.data[:l-1]
    return item
}

func (m *MinHeap[T]) Push(a any) {
    m.data =append(m.data, a.(T))
}

File: 0743-network-delay-time/0743-network-delay-time.rs
use std::collections::BinaryHeap;
use std::cmp::Ordering;

impl Solution {
    
    pub fn network_delay_time(times: Vec<Vec<i32>>, n: i32, k: i32) -> i32 {
        // calculate the shortest path from the source to all nodes (can use djikstra)
        // then calculate the max of dists
        let n = n as usize;
        let k = k as usize;

        let mut adjList: Vec<Vec<Vertex>> = vec![Vec::<Vertex>::new();n as usize];
        for t in times.iter() {
            let (src, dst, cost) = ((t[0]-1) as usize, (t[1]-1) as usize, t[2]);
            adjList[src].push(Vertex{node: dst, cost: cost});
        }
        let dists = Solution::djikstra(k-1, adjList);
        

        let &max_time = dists.iter().max().unwrap();
        if max_time == i32::MAX {
            return -1
        }
        return max_time;
    }

    pub fn djikstra(src: usize, adjList: Vec<Vec<Vertex>>) -> Vec<i32> {
        let n = adjList.len();

        let mut dist = vec![i32::MAX;n];
        dist[src] = 0;

        let mut visited = vec![false;n];

        let mut minHeap = BinaryHeap::new();
        minHeap.push(Vertex{node: src, cost: 0});

        while minHeap.len() > 0 {
            let u = minHeap.pop().unwrap();
            if visited[u.node] {
                continue;
            }

            visited[u.node] = true;

            for &edge in adjList[u.node].iter() {
                if !visited[edge.node] && dist[edge.node] > (dist[u.node] + edge.cost) {
                    dist[edge.node] = dist[u.node] + edge.cost;
                    minHeap.push(Vertex{
                        node: edge.node,
                        cost: dist[edge.node],
                    })
                }
            }

        }

        return dist;
    }

}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct Vertex {
    node: usize,
    cost: i32,
}

impl PartialOrd for Vertex {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        return Some(self.cmp(other));
    }
}

impl Ord for Vertex {
    fn cmp(&self, other: &Self) -> Ordering {
        // we want to prioritize the minimum cost
        // since binaryHeap in rust is maxHeap, we should reverse it
        if self.cost == other.cost {
            return other.node.cmp(&self.node)
        }
        return other.cost.cmp(&self.cost);
    }
}